File: src/View/QuestionPanel.java
package View;

import Model.Answer;
import Model.Question;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Enumeration;
import java.util.List;
import java.util.Objects;

public class QuestionPanel extends JPanel implements ActionListener {

    private Question question;
    private ButtonGroup group;
    private int questionNo;
    private QuestionCallbackEvents mQuestionCallbackEvents;

    QuestionPanel(Question question, int questionNo) {
        this.question = question;
        this.questionNo = questionNo;
        setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
        setBackground(Color.white);

//        JLabel mQuestionLabel = new JLabel("<html>Q" + (questionNo + 1) + ". " + question.getMQuestion()+"</html>");
//        mQuestionLabel.validate();
//        mQuestionLabel.setOpaque(true);
//        mQuestionLabel.setFont(new Font("Calibri",Font.BOLD,22));
//        mQuestionLabel.setBackground(Color.white);
//        add(mQuestionLabel);

        JTextArea mQuestionLabel = new JTextArea("Q" + (questionNo + 1) + ". " + question.getMQuestion());
        mQuestionLabel.setEditable(false);
        mQuestionLabel.setLineWrap(true);
        mQuestionLabel.setWrapStyleWord(true);
        mQuestionLabel.setOpaque(true);
        mQuestionLabel.setFont(new Font("Calibri", Font.BOLD, 22));
        mQuestionLabel.setBackground(Color.white);

        JScrollPane pane = new JScrollPane(mQuestionLabel);
        pane.setBounds(0, 0, 795, 350);
        add(pane);

        List<Answer> choices = question.getMOptions();

        group = new ButtonGroup();
        JRadioButton[] options = new JRadioButton[4];
        for (int i = 0; i < 4; i++) {
            options[i] = new JRadioButton();
            options[i].setBackground(Color.white);
            options[i].setFont(new Font("Calibri", Font.PLAIN, 18));
            options[i].setText(choices.get(i).getMOption());
            options[i].addActionListener(this);
            group.add(options[i]);
            if (isAnswered()) {
                if (Objects.requireNonNull(question.getMMarkedAnswer()).getMId() == choices.get(i).getMId()) {
                    options[i].setSelected(true);
                }
            }
            add(options[i]);
        }

    }

    void setmQuestionCallbackEvents(QuestionCallbackEvents mQuestionCallbackEvents) {
        this.mQuestionCallbackEvents = mQuestionCallbackEvents;
    }

    private boolean isAnswered() {
        return question.getMStatus() == Question.ANSWERED;
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        update();
    }

    void update() {
        int selectedAnswer = getSelectedAnswer(getSelectedButtonText(group));
        if (selectedAnswer == -1)
            return;
        question.setMStatus(Question.ANSWERED);
        question.setMMarkedAnswer(fromOption(selectedAnswer));
        mQuestionCallbackEvents.update(questionNo, question);
    }

    private int getSelectedAnswer(String choiceText) {
        List<Answer> list = question.getMOptions();
        for (Answer ans : list) {
            if (ans.getMOption().equals(choiceText)) {
                return ans.getMId();
            }
        }
        return -1;
    }

    private String getSelectedButtonText(ButtonGroup grp) {
        for (Enumeration<AbstractButton> buttons = grp.getElements(); buttons.hasMoreElements(); ) {
            AbstractButton button = buttons.nextElement();
            if (button.isSelected()) {
                return button.getText();
            }
        }
        return null;
    }

    private Answer fromOption(int option) {
        for (Answer answer : question.getMOptions()) {
            if (answer.getMId() == option)
                return answer;
        }
        return null;
    }
}

File: src/View/QuestionPanel.kt
package View

import Model.Question
import sun.reflect.generics.reflectiveObjects.NotImplementedException
import java.awt.Color
import java.awt.Font
import javax.swing.*

class QuestionPanel(private val mQuestionList: List<Question>) : JPanel() {
    private val mQuestionLabel: JLabel
    private val opt1: JRadioButton
    private val opt2: JRadioButton
    private val opt3: JRadioButton
    private val opt4: JRadioButton
    private var currentQuestion: Int

    init {
        layout = BoxLayout(this, BoxLayout.Y_AXIS)
        background = Color.white
        val grp = ButtonGroup()
        opt1 = JRadioButton()
        opt2 = JRadioButton()
        opt3 = JRadioButton()
        opt4 = JRadioButton()
        opt1.background = Color.white
        opt2.background = Color.white
        opt3.background = Color.white
        opt4.background = Color.white
        opt1.font = Font("Calibri", Font.PLAIN, 18)
        opt2.font = Font("Calibri", Font.PLAIN, 18)
        opt3.font = Font("Calibri", Font.PLAIN, 18)
        opt4.font = Font("Calibri", Font.PLAIN, 18)
        grp.add(opt1)
        grp.add(opt2)
        grp.add(opt3)
        grp.add(opt4)

        currentQuestion = 0

        mQuestionLabel = JLabel()
        mQuestionLabel.font = Font("Serif", Font.BOLD, 28)
        mQuestionLabel.alignmentX = 0.1F
        add(mQuestionLabel)
        add(opt1)
        add(opt2)
        add(opt3)
        add(opt4)

        seek(currentQuestion)

    }

    private fun display(question: Question) {
        mQuestionLabel.text = "Q " + (currentQuestion + 1) + ".  " + question.mQuestion
        val answerList = question.randomizeAnswers()
        opt1.text = answerList[0].mOption
        opt2.text = answerList[1].mOption
        opt3.text = answerList[2].mOption
        opt4.text = answerList[3].mOption
        if (question.mMarkedAnswer == null) {
            updateUI()
            return
        }
        when (question.mMarkedAnswer) {
            answerList[0] -> opt1.isSelected = true
            answerList[1] -> opt2.isSelected = true
            answerList[2] -> opt3.isSelected = true
            answerList[3] -> opt4.isSelected = true
        }
        updateUI()
    }

    fun seek(i: Int) {
        currentQuestion = i
        display(mQuestionList[i])
    }

    fun update(): Question {
        throw NotImplementedException()
        TODO("INSERT UPDATE METHOD WHICH CAN SAVE THE MARKED ANSWER TO THE EXISTING LIST OF QUESTIONS AND RETURN THE UPDATED QUESTION SO THAT CHANGES CAN BE MADE TO SAVE THE CURRENT INSTANCE")
        return mQuestionList[currentQuestion]
    }

    fun next() {
        if (currentQuestion != mQuestionList.size - 1) {
            currentQuestion += 1
            seek(currentQuestion)
        }
    }

    fun prev() {
        if (currentQuestion != 0) {
            currentQuestion -= 1
            seek(currentQuestion)
        }
    }
}